### PR TITLE
fix(hmac-auth): when the request contains escape characters, the signature verification fails

### LINKED ISSUE #2710: bug(hmac-auth): when the request contains escape characters, the signature verification fails
### Issue description

In different programming language environments, the escape characters of parameters in `uri` have the difference between uppercase and lowercase letters, and they all represent the same characters. In the `hmac-auth` plugin, the parameters in `uri` are uniformly escaped with `escape_uri`, which will convert lowercase letters in escaped characters to uppercase letters. Eventually, the signature in the request will be inconsistent with the signature generated by the `hmac-auth` plugin, and the signature verification will fail.

I think a feasible solution: add an `enable_encode` (Boolean type) field in `schema` to control whether to enable `uri` escape. In order to maintain backward compatibility, the `escape_uri` method is used by default (`enable_encode` is true). When `enable_encode` is false, the `escape_uri` is not used.
What do you think? Welcome to discuss together.

### Environment

* apisix version (cmd: `apisix version`): 2.0
* OS:

Example:
1.Add a line of logging code below line 204 of the `hmac-auth` plugin to print the characters escaped by `escape_uri`.

```
core.log.info("query: ", escape_uri(key) .. "=" .. escape_uri(param))
```

截屏2020-11-11 下午5 19 10

2.Add test case with lowercase letter escape characters in the `hmac-auth.t` file.

```
=== TEST 35: test escape_uri
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/routes/1',
                ngx.HTTP_PUT,
                [[{
                    "plugins": {
                        "hmac-auth": {}
                    },
                    "upstream": {
                        "nodes": {
                            "127.0.0.1:1980": 1
                        },
                        "type": "roundrobin"
                    },
                    "uri": "/hello"
                }]]
                )

            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]



=== TEST 36: test escape_uri
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/consumers',
                ngx.HTTP_PUT,
                [[{
                    "username": "james",
                    "plugins": {
                        "hmac-auth": {
                            "access_key": "my-access-key4",
                            "secret_key": "my-secret-key4"                           
                        }
                    }
                }]]
                )
            
            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]



=== TEST 37: The escape characters `%3e` and `%2c` with lowercase letters in the uri parameter
--- config
location /t {
    content_by_lua_block {
        local ngx_time = ngx.time
        local ngx_http_time = ngx.http_time
        local core = require("apisix.core")
        local t = require("lib.test_admin")
        local hmac = require("resty.hmac")
        local ngx_re = require("ngx.re")
        local ngx_encode_base64 = ngx.encode_base64

        local data = {cert = "ssl_cert", key = "ssl_key", sni = "test.com"}
        local req_body = core.json.encode(data)
        req_body = req_body or ""

        local secret_key = "my-secret-key4"
        local timestamp = ngx_time()
        local gmt = ngx_http_time(timestamp)
        local access_key = "my-access-key4"
        local custom_header_a = "asld$%dfasf"
        local custom_header_b = "23879fmsldfk"

        local signing_string = {
            "GET",
            "/hello",
            "name=rose%3ehello&name2=james%2chello",
            access_key,
            gmt,
            "x-custom-header-a:" .. custom_header_a,
            "x-custom-header-b:" .. custom_header_b
        }
        signing_string = core.table.concat(signing_string, "\n") .. "\n"
        core.log.info("signing_string:", signing_string)

        local signature = hmac:new(secret_key, hmac.ALGOS.SHA256):final(signing_string)
        core.log.info("signature:", ngx_encode_base64(signature))
        local headers = {}
        headers["X-HMAC-SIGNATURE"] = ngx_encode_base64(signature)
        headers["X-HMAC-ALGORITHM"] = "hmac-sha256"
        headers["Date"] = gmt
        headers["X-HMAC-ACCESS-KEY"] = access_key
        headers["X-HMAC-SIGNED-HEADERS"] = "x-custom-header-a;x-custom-header-b"
        headers["x-custom-header-a"] = custom_header_a
        headers["x-custom-header-b"] = custom_header_b

        local code, body = t.test('/hello?name=rose%3ehello&name2=james%2chello',
            ngx.HTTP_GET,
            req_body,
            nil,
            headers
        )

        if code >= 300 then
            ngx.status = code
        end
        ngx.say(body)           
    }
}
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]
```

3.Test case execution result

截屏2020-11-11 下午6 02 39

View log：

截屏2020-11-11 下午6 06 34

### PR DESCRIPTION
**Pull request by Firstsawyou**
fix #2710

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: doc/zh-cn/plugins/hmac-auth.md [hunk 1]
```diff
@@ -44,6 +44,7 @@
 | clock_skew     | integer       | 可选   | 0           |                                             | 签名允许的时间偏移，以秒为单位的计时。比如允许时间偏移 10 秒钟，那么就应设置为 `10`。特别地，`0` 表示不对 `Date` 进行检查。                                                        |
 | signed_headers | array[string] | 可选   |               |                                             | 限制加入加密计算的 headers ，指定后客户端请求只能在此范围内指定 headers ，此项为空时将把所有客户端请求指定的 headers 加入加密计算。如： ["User-Agent", "Accept-Language", "x-custom-a"] |
 | keep_headers | boolean | 可选   |      false        |           [ true, false ]                             | 认证成功后的 http 请求中是否需要保留 `X-HMAC-SIGNATURE`、`X-HMAC-ALGORITHM` 和 `X-HMAC-SIGNED-HEADERS` 的请求头。true: 表示保留 http 请求头，false: 表示移除 http 请求头。 |
+| enable_encode | boolean | 可选   |      true        |           [ true, false ]                             | 是否对签名中的 uri 参数进行转义,例如: `params1=hello%2Cworld` 进行了转义，`params2=hello,world` 没有进行转义。true: 表示对签名中的 uri 参数进行转义，false: 不对签名中的 uri 参数转义。 |
```
Comment (spacewander) [inline doc/zh-cn/plugins/hmac-auth.md:47]: Better to update the example in `测试插件`

File: apisix/plugins/hmac-auth.lua [hunk 2]
```diff
@@ -194,14 +209,28 @@ local function generate_signature(ctx, secret_key, params)
...
+        local field_val = get_conf_field(params.access_key, "enable_encode")
+        core.log.info("enable_encode: ", field_val)
+
         for _, key in pairs(keys) do
             local param = args[key]
-            if type(param) == "table" then
-                for _, val in pairs(param) do
-                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(val))
+            -- whether to escape the uri parameters
+            if field_val then
```
Comment (spacewander) [inline apisix/plugins/hmac-auth.lua:218]: Would be better if we use an empty function to escape the uri params when the option is disable. Like, `core.table.insert(query_tab, encode_or_not(key) .. "=" .. encode_or_not(val))`. When you do it, note that, please avoid creating a new function per request.

File: apisix/plugins/hmac-auth.lua [hunk 3]
```diff
@@ -74,6 +74,11 @@ local consumer_schema = {
             type = "boolean",
             title = "whether to keep the http request header",
             default = false,
+        },
+        enable_encode = {
```
Comment (spacewander) [inline apisix/plugins/hmac-auth.lua:78]: Better to use `encode_uri_param` so people can know what is encoded.

File: apisix/plugins/hmac-auth.lua (hunk 2 above)
Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua:218]: @spacewander 
You mean we need to use a function like this: `local function encode_or_not(key) return key end` . To handle the situation where this option is disabled?

Comment (spacewander) [reply apisix/plugins/hmac-auth.lua:218]: `encode_or_not` is a variable, like that: `if xxx then encode_or_not = encode else encode_or_not = do_nothing`.

File: doc/zh-cn/plugins/hmac-auth.md (hunk 1 above)
Comment (Firstsawyou) [reply doc/zh-cn/plugins/hmac-auth.md]: added.

File: apisix/plugins/hmac-auth.lua (hunk 3 above)
Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua]: done.

File: apisix/plugins/hmac-auth.lua (hunk 2 above)
Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua]: @spacewander  PTAL. I am not sure if the modification here is correct.

Comment (spacewander) [reply apisix/plugins/hmac-auth.lua]: Actually I mean this change:
```diff
diff --git apisix/plugins/hmac-auth.lua apisix/plugins/hmac-auth.lua
index bc09d21..45cc0cc 100644
--- apisix/plugins/hmac-auth.lua
+++ apisix/plugins/hmac-auth.lua
@@ -190,6 +190,11 @@ local function get_conf_field(access_key, field_name)
 end


+local function do_nothing(v)
+    return v
+end
+
+
 local function generate_signature(ctx, secret_key, params)
     local canonical_uri = ctx.var.uri
     local canonical_query_string = ""
@@ -212,25 +217,22 @@ local function generate_signature(ctx, secret_key, params)
         local encode_or_not = get_conf_field(params.access_key, "encode_uri_params")
         core.log.info("encode_uri_params: ", encode_or_not)

+        local encode_uri_param = do_nothing
+        if encode_or_not then
+            encode_uri_param = escape_uri
+        end
+
         for _, key in pairs(keys) do
             local param = args[key]
             -- whether to encode the uri parameters
-            if encode_or_not then
-                if type(param) == "table" then
-                    for _, val in pairs(param) do
-                        core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(val))
-                    end
-                else
-                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(param))
+            if type(param) == "table" then
+                for _, val in pairs(param) do
+                    core.table.insert(query_tab, encode_uri_param(key) .. "="
+                                      .. encode_uri_param(val))
                 end
             else
-                if type(param) == "table" then
-                    for _, val in pairs(param) do
-                        core.table.insert(query_tab, key .. "=" .. val)
-                    end
-                else
-                    core.table.insert(query_tab, key .. "=" .. param)
-                end
+                core.table.insert(query_tab, encode_uri_param(key) .. "="
+                                  .. encode_uri_param(param))
             end
         end
         canonical_query_string = core.table.concat(query_tab, "&")
```

You can integrate it into your code.

Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua]: ok, thank you very much.

**Review by spacewander — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Pull request merged**